File: library/kotlin/io/envoyproxy/envoymobile/stats/GaugeImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Gauge`.
 */
internal class GaugeImpl : Gauge {
  var envoyEngine: WeakReference<EnvoyEngine>
  var series: String
  var tags: Map<String, String>

  internal constructor(engine: EnvoyEngine, elements: List<Element>, tags: List<Tag>) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.series = elements.joinToString(separator = ".") { it.value }
    this.tags = convert(tags)
  }

  private constructor(engine: WeakReference<EnvoyEngine>, series: String, tags: List<Tag>) {
    this.envoyEngine = engine
    this.series = series
    this.tags = convert(tags)
  }

  override fun set(value: Int) {
    envoyEngine.get()?.recordGaugeSet(series, tags, value)
  }

  override fun add(amount: Int) {
    envoyEngine.get()?.recordGaugeAdd(series, tags, amount)
  }

  override fun sub(amount: Int) {
    envoyEngine.get()?.recordGaugeSub(series, tags, amount)
  }

  override fun attach(tags: List<Tag>): Gauge {
    return GaugeImpl(envoyEngine, series, tags)
  }
}

File: library/swift/stats/GaugeImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of time series gauge.
@objcMembers
final class GaugeImpl: NSObject, Gauge {
  private let series: String
  private let tags: [String: String]
  private weak var engine: EnvoyEngine?

  init(elements: [Element], tags: [Tag], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    self.tags = Stats.convert(tags: tags)
    super.init()
  }

  init(series: String, tags: [Tag], engine: EnvoyEngine?) {
    self.series = series
    self.engine = engine
    self.tags = Stats.convert(tags: tags)
    super.init()
  }

  /// Set the gauge with the given value.
  func set(value: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordGaugeSet(self.series, tags: self.tags, value: numericCast(value))
  }

  /// Add the given amount to the gauge.
  func add(amount: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordGaugeAdd(self.series, tags: self.tags, amount: numericCast(amount))
  }

  /// Subtract the given amount from the gauge.
  func sub(amount: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordGaugeSub(self.series, tags: self.tags, amount: numericCast(amount))
  }

  func attach(tags: [Tag]) -> Gauge {
    return GaugeImpl(series: self.series, tags: tags, engine: self.engine)
  }
}
